test/test_torch.py

```diff
@@ -4777,23 +4844,92 @@ def import_module(name, filename):
                     self.assertEquals(len(w), 1)
                     self.assertTrue(w[0].category, 'SourceChangeWarning')
 
+    def test_serialization_container(self):
+        self._test_serialization_container(lambda: tempfile.NamedTemporaryFile())
+
+    # TODO(rzou): test_serialization_container breaks if called multiple times.
+    # The following test won't fail if called by itself.
+    @unittest.skipIf(not PY3, 'This test is broken on Python 2')
+    def test_serialization_container_filelike(self):
+        self._test_serialization_container(lambda: BytesIOContext())
+
     def test_serialization_map_location(self):
         test_file_path = download_file('https://download.pytorch.org/test_data/gpu_tensors.pt')
 
         def map_location(storage, loc):
             return storage
 
-        tensor = torch.load(test_file_path, map_location=map_location)
-        self.assertIsInstance(tensor, torch.FloatTensor)
-        self.assertEqual(tensor, torch.FloatTensor([[1.0, 2.0], [3.0, 4.0]]))
+        def load_bytes():
+            with open(test_file_path, 'rb') as f:
+                data = io.BytesIO(f.read())
+            return data
+
+        fileobject_lambdas = [lambda: test_file_path, load_bytes]
+        map_locations = [map_location, {'cuda:0': 'cpu'}, 'cpu']
+
+        for fileobject_lambda in fileobject_lambdas:
+            for map_location in map_locations:
+                tensor = torch.load(fileobject_lambda(), map_location=map_location)
+                self.assertIsInstance(tensor, torch.FloatTensor)
+                self.assertEqual(tensor, torch.FloatTensor([[1.0, 2.0], [3.0, 4.0]]))
+
+    def test_serialization_api_guarantees(self):
+        filemock = FilelikeMock(b'', has_readinto=False)
+        tensor = torch.randn(3, 5)
+        torch.save(tensor, filemock)
+        expected = set(['write', 'flush'])
+        self.assertTrue(len(filemock.calls.difference(expected)) is 0)
+
+        # Reset between save and load
+        filemock.seek(0)
+        filemock.calls.clear()
+
+        _ = torch.load(filemock)
+        expected = set(['read', 'readline', 'seek', 'tell'])
```

Comment:
do we really "guarantee" that we will use tell and seek? That doesn't look like something we want to test

Reply:
I was trying to test that when someone calls torch.read(), it'll make a subset of (read, readline, seek, and tell) calls, and call nothing outside of those functions (because those are what I put into the documentation). I'll make this clearer